shared/health/src/main/java/io/pravega/shared/health/impl/HealthEndpointImpl.java

```diff
@@ -0,0 +1,136 @@
+/**
+ * Copyright Pravega Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.pravega.shared.health.impl;
+
+import io.pravega.shared.health.Health;
+import io.pravega.shared.health.HealthContributor;
+import io.pravega.shared.health.HealthEndpoint;
+import io.pravega.shared.health.Status;
+import io.pravega.shared.health.HealthServiceUpdater;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Provides a simpler interface for retrieving {@link Health} information from the {@link io.pravega.shared.health.HealthService}.
+ * The {@link HealthEndpoint} should be used to serve information about the {@link Health} of a component for external classes.
+ */
+@Slf4j
+@RequiredArgsConstructor
+public class HealthEndpointImpl implements HealthEndpoint {
+
+    @NonNull
+    private final HealthContributor root;
```

Comment:
`HealthEndpointImpl` and `HealthServiceImpl` look identical to me. Both hold the root HealthContributor object and the updater object. Both are also exposed to clients.
How are these 2 APIs different: `healthService.getHealth()` and  `healthService.getEndpoint().getHealth()` ? You may consider refactoring these into a single class.

Reply:
I see your point -- there definitely is a little overlap but I wouldn't consider them __identical__ per se. There is no `healthService.getHealth()` call and the `HealthServiceUpdater` and root `HealthContriubtor` are not directly exposed in the `HealthEndpoint` as they are in the `HealthService`. I think there is value in separating the current `HealthEndpoint` logic in it's own class but I don't feel strongly one way or the other. @andreipaduroiu Do you have any thoughts on this?

Reply:
I looked at both these classes and I see both your points. The way these will be used, from what I understand, is that the application bootstrapper (`ServiceStarter` for segment store and `ControllerServiceStarter` for the controller) will create a `HealthServiceImpl` object and manage it, while the soon-to-be-added REST health endpoint will call `HealthService.getEndpoint()` and work with that.

So in essence, the `HealthServiceImpl` is nothing more than a manager for `HealthEndpoint`. Another term we could use to describe it is a "factory", however, it does not create new `HealthEndpoint` objects, it only has one and it always returns that one.

From another POV, the `HealthServiceImpl` adds some value on top of `HealthEndpointImpl`, but not much. I can see why we'd want to merge the two together. However, from a testing POV, it will be easier to keep the separation since we can mock a `HealthEndpoint` without having to worry about whatever else the `HealthService` might do.

My recommendation, to simplify the design would be:
1. Remove the `HealthService` interface (whatever `HealthServiceImpl` implements). We only expect one of these objects per process and there's no point in maintaining an interface as well.
2. Rename `HealthServiceImpl` to `HealthEndpointManager`. Keep all logic within it the same.
3. Consider what exactly the newly named `HealthEndpointManager` should expose. It does need to expose the endpoint, but does it need to expose the root or updater? If the answer is "only for testing", then add package-private getters and mark them as `@VisibleForTesting`; if the answer is "no", then don't expose them at all.

Reply:
Yes this sounds good to me.